### DISCUSSION TITLE
Subpixel rendering and very small font sizes

### DISCUSSION THREAD

**Discussion by JackiBackiBoy** [Chlumsky/msdf-atlas-gen#69] [General]
Hi again! The MSDF atlas generation you've created is absolutely gorgeous, and is offering very good rendering capabilities for both small and larger font sizes. However, I have noticed that for very small font sizes (12pt and below) that the antialiasing, while very good, is not always as clear as text rendered with the subpixel rendering technique. The main difference is that the antialiasing at very small font sizes acts more like a **blur**. I am aware that Freetype has subpixel rendering as an option, but I am wondering if msdf-atlas-gen can be used to construct subpixel rendered text?

Might it be possible to take an MSDF rendered font atlas and translate it into its subpixel rendered equivalent? If not, does msdf-atlas-gen offer texture atlases generated with subpixel rendering, and might there be other techniques to increase readability of small text? I would really appreciate your input on the matter.

Thank you in advance!

**Comment by saidwho12** [comment]
I've done it before it's possible to do with OpenGL look into "Dual Source Blending" and OpenGL blending functions in general and how subpixel AA works. It's gonna work differently based on the display's subpixel geometry.

**Comment by Chlumsky** [comment]
Before you start implementing some complicated solution, make sure that your SDF's pixel range is high enough. The default value is only sufficient for rendering at the same size as the SDF or larger.

But again, subpixel rendering has nothing to do with MSDF. You can subpixel-render anything and the way you do it doesn't change. You could make a whole game engine that renders everyhing with subpixel anti-aliasing. However, you need to know the layout of the pixels of the user's monitor and that it is running in its native resolution. This is why subpixel text rendering is usually handled by the OS, which has this information. But if you do know the monitor's pixel layout, you just draw each color channel with a slightly offset position based on the layout and that's all there is to it.

As for the other techniques, you might want to try supersampling if the SDF is being significantly downscaled, but the primary solution for this problem in text rendering is hinting, which is not really a good fit for SDF-based rendering.

**Reply by JackiBackiBoy** [reply]
Regarding the SDF pixel range, is there a recommended way to calculate it? Or is it just a balancing act? For some font sizes a pixel range of 2.0 seems fine, for other 3.5 has worked great.

**Reply by Chlumsky** [reply]
2/scale

**Reply by Chlumsky** [reply]
Actually 1+1/scale might be enough.

**Reply by Chlumsky** [reply]
Alright, I just tested it and the minimum value that is guaranteed to work is actually 2.5+1/scale. I guess my default setting is completely wrong then.

**Reply by JackiBackiBoy** [reply]
Alright, thank you for the clarification. Is the `scale` you're referring to the minimum scale specified when calling `packer.setMinimumScale(scale)`? Or should the pixel range be continously updated when passing it to the shader?

**Reply by Chlumsky** [reply]
No, I'm talking about scale at which you're drawing. Pixel range is an immutable property of the generated SDF.

**Reply by JackiBackiBoy** [reply]
Yes! I am now getting much better rendering results. Thank you for the assistance, much appreciated!
